File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-9.cs
public class Solution {
    // An optimal solution would be to use a two pointer algorithm where every substring is represented
    // as a window that either expands when no duplicates are found, or shrinks when a duplicate is found.
    // A hashSet would be used at every index to keep track of duplicate characters.
    // And the max lenght would be stored 
    // Time Complexity would be O(n) since we're visiting every character at most once
    // Space Complexity would be O(m) where m is the number of unique characters in the array.

    public int LengthOfLongestSubstring(string s) {
        var hashSet = new HashSet<char>();
        var length = 0;
        var l = 0;

        for (var r = 0; r < s.Length; r++) {
            while (hashSet.Contains(s[r])) {
                hashSet.Remove(s[l]);
                l++;        
            };
            
            hashSet.Add(s[r]);
            length = Math.Max(hashSet.Count, length);
        }
        

        return length;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-9.ts
class Solution {
    /**
     * @param {string} s
     * @param {number} k
     * @return {number}
     */
    characterReplacement(s: string, k: number): number {
        const hashMap = new Map();
        let res = 0;
        let l = 0;
        let r = 0;

        while (r < s.length) {
            hashMap.set(s[r], (hashMap.get(s[r]) || 0) + 1);

            // is window is valid
            while ((r - l + 1 - Math.max(...hashMap.values())) > k) {
                if (hashMap.get(s[l]) > 0) hashMap.set(s[l], hashMap.get(s[l]) - 1)
                l++
            }

            console.log(hashMap, res)
            res = Math.max(res, r - l + 1)
            r++
        }

        return res
    }
}
